### ISSUE TITLE
Plugin loading on plugins_loaded at PHP_INT_MIN before other plugins properly initialize

### ISSUE THREAD

**Issue by rmccue** [Automattic/amp-wp#6499]
## Bug Description

AMP loads at PHP_INT_MIN:

https://github.com/ampproject/amp-wp/blob/890617a5f28e555a63b4b28074a7dd949f52dec1/amp.php#L259

This means that other plugins may not properly have initialized, including dependencies of AMP itself. 

In particular, AMP calls `wp_upload_dir()` via:
```
amp_bootstrap_plugin
AmpProject\AmpWP\Infrastructure\ServiceBasedPlugin->register
AmpProject\AmpWP\Infrastructure\ServiceBasedPlugin->register_services
AmpProject\AmpWP\Infrastructure\ServiceBasedPlugin->register_service
AmpProject\AmpWP\Infrastructure\ServiceBasedPlugin->instantiate_service
AmpProject\AmpWP\Infrastructure\Injector\SimpleInjector->make
AmpProject\AmpWP\Infrastructure\Injector\SimpleInjector->get_dependencies_for
array_map
AmpProject\AmpWP\Infrastructure\Injector\SimpleInjector->AmpProject\AmpWP\Infrastructure\Injector\{closure}
AmpProject\AmpWP\Infrastructure\Injector\SimpleInjector->resolve_argument
AmpProject\AmpWP\Infrastructure\Injector\SimpleInjector->make_dependency
AmpProject\AmpWP\Infrastructure\Injector\SimpleInjector->get_dependencies_for
array_map
AmpProject\AmpWP\Infrastructure\Injector\SimpleInjector->AmpProject\AmpWP\Infrastructure\Injector\{closure}
AmpProject\AmpWP\Infrastructure\Injector\SimpleInjector->resolve_argument
AmpProject\AmpWP\Infrastructure\Injector\SimpleInjector->make_dependency
AmpProject\AmpWP\Infrastructure\Injector\FallbackInstantiator->instantiate
AmpProject\Optimizer\TransformationEngine->__construct
AmpProject\Optimizer\TransformationEngine->initializeTransformers
AmpProject\Optimizer\TransformationEngine->getTransformerDependencies
AmpProject\AmpWP\Optimizer\AmpWPConfiguration->getTransformerConfiguration
AmpProject\Optimizer\DefaultConfiguration->getTransformerConfiguration
AmpProject\Optimizer\Configuration\BaseTransformerConfiguration->__construct
AmpProject\AmpWP\Optimizer\Transformer\AmpSchemaOrgMetadataConfiguration->getAllowedKeys
amp_get_schemaorg_metadata
amp_get_publisher_logo
wp_get_attachment_image_src
image_downsize
wp_get_attachment_url
wp_get_upload_dir
wp_upload_dir
```

This means that any plugin which changes the upload directory, such as [S3 Uploads](https://github.com/humanmade/S3-Uploads), has not been registered at this point. This causes the wrong URLs and paths to be used, as well as causing an invalid value to be cached in Altis (https://github.com/humanmade/altis-cloud/issues/421).

## Expected Behaviour

AMP should only initialize and add hooks at a very early stage, not actually collect data from WordPress. (Also, do you _really_ need to use PHP_INT_MAX?)

## Steps to reproduce

(n/a, but I can provide a small replication of S3 Uploads' code if absolutely necessarily)

## Additional context

 - WordPress version: n/a
 - Plugin version: Latest (original uploads bug in 2.1.1, but early plugins_loaded priority is in latest)
 - Gutenberg plugin version (if applicable):
 - AMP plugin template mode:
 - PHP version:
 - OS:
 - Browser: [e.g. chrome, safari]

---------------

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Acceptance criteria

* 

## Implementation brief

* 

## QA testing instructions

* 

## Demo

* 

## Changelog entry

*

**Comment by rmccue** [comment]
This may _actually_ have been fixed in 02ace15e42a78eec1a256cd65a2765a4c0c72bbe, investigating; the version doesn't match exactly what I saw initially. (Honestly, trying to track these 40 levels of indirection is doing my head in a bit, so apologies if it has been.)

**Comment by rmccue** [comment]
This appears to have been fixed as a side-effect of 02ace15e42a78eec1a256cd65a2765a4c0c72bbe, so I _think_ the original bug surfaced by this issue has been fixed.

That said, loading the plugin this early does mess with other systems and can lead to unintended behaviour, so I'll leave this open at your discretion.

**Comment by westonruter** [comment]
> This appears to have been fixed as a side-effect of [02ace15](https://github.com/ampproject/amp-wp/commit/02ace15e42a78eec1a256cd65a2765a4c0c72bbe), so I _think_ the original bug surfaced by this issue has been fixed.

The PR with this fix was #6233. You can see with that change the `upload_dir` filter is no longer applying, if I understand the problem correctly with your conflict.

> That said, loading the plugin this early does mess with other systems and can lead to unintended behaviour, so I'll leave this open at your discretion.

It could have called `amp_bootstrap_plugin()` right there in `amp.php` without deferring it to run at `plugins_loaded`. So in that way, it could actually be running earlier.

> AMP should only initialize and add hooks at a very early stage, not actually collect data from WordPress. (Also, do you _really_ need to use PHP_INT_MAX?)

Are you actually referring to `PHP_INT_MIN` for the `plugins_loaded` action at which `amp_bootstrap_plugin()` is run? The reason why we wait for `plugins_loaded` has to do with the Plugin Suppression functionality. We need all plugins to have been loaded with their own hooks registered so that we can unregister them if a given plugin has been suppressed from running on AMP pages. At least, this was the intention from 445f4554b4b1b2ba7c277ac9c517e4cabc1f8e58. But then later this was deferred to a later priority in 2d4dea46439bc562c2878a28ceec493cfed62980, so that isn't the necessarily a reason anymore. 

So we could postpone to calling `amp_bootstrap_plugin()` to a later point (e.g. `plugins_loaded` with priority 0) if it is currently still causing a problem for you, which it appears it is not.
